fastestimator/estimator/trace.py

```diff
@@ -195,6 +195,97 @@ def on_epoch_end(self, state):
 
     def on_end(self, state):
         state['total_time'] = "{} sec".format(round(time.perf_counter() - self.train_start, 2))
+        for model_name, model in self.network.model.items():
+            state[model_name + "_lr"] = round(backend.get_value(model.optimizer.lr), 6)
+
+
+class LRController(Trace):
+    def __init__(self,
+                 model_name,
+                 lr_schedule=None,
+                 reduce_on_eval=False,
+                 reduce_patience=10,
+                 reduce_factor=0.1,
+                 reduce_mode="min",
+                 min_lr=1e-6):
+        if reduce_on_eval and isinstance(reduce_on_eval, str):
+            super().__init__(inputs=reduce_on_eval)
+        else:
+            super().__init__()
+        self.model_name = model_name
+        self.lr_schedule = lr_schedule
+        self.reduce_on_eval = reduce_on_eval
+        self.reduce_patience = reduce_patience
+        self.reduce_factor = reduce_factor
+        self.reduce_mode = reduce_mode
+        self.min_lr = min_lr
+        self.reduce_lr_ratio = 1.0
+        self.base_lr = None
+        self.current_lr = None
+        self.log_steps = None
+        self.model = None
+        self.change_lr = False
+        self.wait = 0
+        if self.lr_schedule:
+            assert isinstance(self.lr_schedule, LRSchedule), "lr_schedule must be instance of LRSchedule"
+        if self.reduce_mode == "min":
+            self.reduce_metric_best = np.Inf
+        elif self.reduce_mode == "max":
+            self.reduce_metric_best = -np.Inf
+        else:
+            raise ValueError("reduce_mode must be either 'min' or 'max'")
+
+    def on_begin(self, state):
+        self.model = self.network.model[self.model_name]
+        self.base_lr = backend.get_value(self.model.optimizer.lr)
+        self.current_lr = max(self.base_lr * self.reduce_lr_ratio, self.min_lr)
+        if self.reduce_on_eval is True:
+            self.reduce_on_eval = self.model.loss_name
+        if self.lr_schedule:
+            self.lr_schedule.initial_lr = self.current_lr
+
+    def on_epoch_begin(self, state):
+        if self.lr_schedule and self.lr_schedule.schedule_mode == "epoch":
+            self.base_lr = self.lr_schedule.schedule_fn(state["epoch"], self.base_lr)
+            self.change_lr = True
+
+    def on_batch_begin(self, state):
+        if state["mode"] == "train":
+            if self.lr_schedule and self.lr_schedule.schedule_mode == "step":
+                self.base_lr = self.lr_schedule.schedule_fn(state["train_step"], self.base_lr)
+                self.change_lr = True
+            if self.change_lr:
+                self._update_lr()
+
+    def on_batch_end(self, state):
+        if state["mode"] == "train" and self.log_steps and state["train_step"] % self.log_steps == 0:
```

Comment:
is there any reason to only emit this on log_steps? doing this if check might be just as expensive as just always putting it in the dictionary

Reply:
yes, the performance improved when only emit this on log_steps

also, we don't want unnecessary stuff to be included in the state dictionary if they are not going to be be used anywhere.

Reply:
huh. interesting

Reply:
yup